third_party/libunwindstack/PeCoffInterface.cpp

```diff
@@ -0,0 +1,271 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unwindstack/PeCoffInterface.h>
+
+#include <unwindstack/Error.h>
+#include <unwindstack/Log.h>
+#include <unwindstack/Memory.h>
+
+namespace unwindstack {
+
+bool PeCoffMemory::Get8(uint8_t* value) {
+  bool success = memory_->Read8(cur_offset_, value);
+  if (success) {
+    cur_offset_ += sizeof(uint8_t);
+  }
+  return success;
+}
+
+bool PeCoffMemory::Get16(uint16_t* value) {
+  bool success = memory_->Read16(cur_offset_, value);
+  if (success) {
+    cur_offset_ += sizeof(uint16_t);
+  }
+  return success;
+}
+
+bool PeCoffMemory::Get32(uint32_t* value) {
+  bool success = memory_->Read32(cur_offset_, value);
+  if (success) {
+    cur_offset_ += sizeof(uint32_t);
+  }
+  return success;
+}
+
+bool PeCoffMemory::Get64(uint64_t* value) {
+  bool success = memory_->Read64(cur_offset_, value);
+  if (success) {
+    cur_offset_ += sizeof(uint64_t);
+  }
+  return success;
+}
+
+bool PeCoffMemory::GetMax64(uint64_t* value, uint64_t size) {
+  switch (size) {
+    case 1:
+      uint8_t value8;
+      if (!Get8(&value8)) {
+        return false;
+      }
+      *value = value8;
+      return true;
+    case 2:
+      uint16_t value16;
+      if (!Get16(&value16)) {
+        return false;
+      }
+      *value = value16;
+      return true;
+    case 4:
+      uint32_t value32;
+      if (!Get32(&value32)) {
+        return false;
+      }
+      *value = value32;
+      return true;
+    case 8:
+      return Get64(value);
+    default:
+      return false;
+  }
+  return false;
+}
+
+template <typename AddressType>
+bool PeCoffInterface<AddressType>::ParseDosHeader(uint64_t offset) {
+  coff_memory_.set_cur_offset(offset);
+  if (!coff_memory_.Get16(&dos_header_.e_magic)) {
+    last_error_.code = ERROR_MEMORY_INVALID;
+    last_error_.address = coff_memory_.cur_offset();
+    return false;
+  }
+
+  constexpr uint16_t kMsDosTwoPointZeroMagicValue = 0x5a4d;
+  if (dos_header_.e_magic != kMsDosTwoPointZeroMagicValue) {
+    Log::Error("Magic MS-DOS 2.0 value not found. Value read: %x", dos_header_.e_magic);
+    last_error_.code = ERROR_INVALID_COFF;
+    return false;
+  }
+
+  // There is data in the DOS header that we don't need, but we still want to make sure that
+  // we can correctly read the memory at these addresses.
+  constexpr size_t kDosHeaderSize = 0x40;
+  std::vector<uint8_t> unused_data(kDosHeaderSize - sizeof(uint16_t) - sizeof(uint32_t));
+  if (!coff_memory_.ReadFully(coff_memory_.cur_offset(), &unused_data[0], unused_data.size())) {
+    last_error_.code = ERROR_MEMORY_INVALID;
+    last_error_.address = coff_memory_.cur_offset();
+    return false;
+  }
+
+  coff_memory_.set_cur_offset(0x3c);
```

Comment:
If this is something used frequently consider implementing `GetFully`.

Reply:
Yes, you are correct. As far as I know, the initial offset will always be 0, so this will always be 0x3c, but it's for sure inconsistent with the rest. Let's keep it in generality with `offset` as the argument, so as you said this needs to be `offset + 0x3c`.